### ISSUE TITLE
Software Stack for Apps

### ISSUE THREAD

**Issue by rucuriousyet** [lirios/appcenter#11]
I am interested in contributing to this project but would prefer to build apps using Golang and QML/Qt (therecipe/qt is a really good binding I use)... is this an option the team would be open to for certain apps? Golang is very mature/stable, fast, compiles to a static binary and is very easy to write/learn... Go would be a great choice for any non-low level application.

**Comment by plfiorini** [comment]
I don't think that Golang is good enough compared to C++ for Qt based applications because it's not an object oriented language.

Golang is best suited for low-level applications, network application and those that need concurrency.

**Comment by rucuriousyet** [comment]
Golang has has OOP support... It's really not a low level lanuguage like
you say, it's more like a faster/better Java... I would say that is way
better for higher level OS applications as it speeds up Dev time and makes
contribution easier since the tool chain required is built in.

-sent from my OnePlus X

**Comment by rucuriousyet** [comment]
Not to mention, that concurrency you spoke of is great for loading
resources and handling processing on background threads while the UI runs
on the main thread. Makes for a very snappy desktop app. If you'd like, I
could submit a proof of concept app using Go and QML?

-sent from my OnePlus X

**Comment by tim-sueberkrueb** [comment]
@rucuriousyet Go is not a "pure" object-oriented language (in comparison to C++ and Java). It has structs and methods but [it doesn't support inheritance for example](https://golang.org/doc/faq#Is_Go_an_object-oriented_language). Qt however is pretty much based on OOP and type hierarchy. 
Still, those [Qt bindings for Go](https://github.com/therecipe/qt) seem to be pretty popular so one would think they must have found a way to workaround that? Could you elaborate a bit how they do that (e.g. how would you create a "subclass" of a Qt class in Go)?

> If you'd like, I could submit a proof of concept app using Go and QML?

What I'd find pretty interesting would be a Go program exposing some sort of custom `QAbstractListModel` (written in Go) to a QML view.

**Comment by rucuriousyet** [comment]
I can't speak for how Go handles creating native Qt classes as all my projects using Qt/Go have been QML based (components and all). I personally prefer QML anyways as it's quite a bit simpler, and gives me to freedom to swap backends with little effort. Its my understanding that Go and QML would be great for any top-level system apps that don't require perfect performance but are still snappy and efficient.

**Comment by rucuriousyet** [comment]
Go would also be really great for any network utilities, package managers,

**Comment by tim-sueberkrueb** [comment]
> I can't speak for how Go handles creating native Qt classes

@rucuriousyet the thing is: sooner or later you'll have to create a class derived from some Qt type in order to export it to QML. Take custom models for example. Without the ability to define types derived from Qt classes and export them to QML, I can't really see how Go can replace a C++ backend for a QML application.
Sure, for certain use cases it will be enough to have some callbacks to communicate with a Go backend but by far not for all use cases, I think, at least if you want to keep a sensible architecture.
NOTE: As I have no experience with the mentioned Go bindings for Qt, I might be wrong with my assessment!

**Comment by rucuriousyet** [comment]
Ill have to research / talk to some people to find out more, there's most likely a way to do this, I just don't know it yet.

**Comment by tim-sueberkrueb** [comment]
> Go would also be really great for any network utilities, package managers,

Go is certainly nice for that sort of things :)

**Comment by rucuriousyet** [comment]
If some tooling is needed for LiriOS that could be written in Go, let me know! I would definitely be interested in helping out.
